src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+            if (!HasProblem(component))
+            {
+                return TryGetRangeComponentValueWorker(component, out ch);
+            }
+
+            ch = default;
+            return false;
+        }
+
+        private bool TryGetRangeComponentValueWorker(RegexNode component, out char ch)
+        {
+            switch (component.Kind)
+            {
+                case RegexKind.SimpleEscape:
+                    ch = ((RegexSimpleEscapeNode)component).TypeToken.VirtualChars[0];
+                    switch (ch)
+                    {
+                        case 'a': ch = '\u0007'; break;
+                        case 'b': ch =  '\b'; break;
+                        case 'e': ch = '\u001B'; break;
+                        case 'f': ch = '\f'; break;
+                        case 'n': ch = '\n'; break;
+                        case 'r': ch = '\r'; break;
+                        case 't': ch = '\t'; break;
+                        case 'v': ch = '\u000B'; break;
+                    }
+                    return true;
+
+                case RegexKind.ControlEscape:
+                    var controlEscape = (RegexControlEscapeNode)component;
+                    var controlCh = controlEscape.ControlToken.VirtualChars[0].Char;
+                    // \ca interpreted as \cA
+                    if (controlCh >= 'a' && controlCh <= 'z')
+                    {
+                        controlCh -= (char)('a' - 'A');
+                    }
+                    ch = (char)(controlCh - '@');
+                    return true;
+
+                case RegexKind.OctalEscape:
+                    ch = GetCharValue(((RegexOctalEscapeNode)component).OctalText, withBase: 8);
+                    return true;
+
+                case RegexKind.HexEscape:
+                    ch = GetCharValue(((RegexHexEscapeNode)component).HexText, withBase: 16);
+                    return true;
+
+                case RegexKind.UnicodeEscape:
+                    ch = GetCharValue(((RegexUnicodeEscapeNode)component).HexText, withBase: 16);
+                    return true;
+
+                case RegexKind.PosixProperty:
+                    // When the native parser sees [:...:] it treats this as if it just saw '[' and skipped the 
+                    // rest.
+                    ch = '[';
+                    return true;
+
+                case RegexKind.Text:
+                    ch = ((RegexTextNode)component).TextToken.VirtualChars[0];
+                    return true;
+
+                case RegexKind.Sequence:
+                    var sequence = (RegexSequenceNode)component;
+#if DEBUG
+                    Debug.Assert(sequence.ChildCount > 0);
+                    for (int i = 0, n = sequence.ChildCount - 1; i < n; i++)
+                    {
+                        Debug.Assert(IsEscapedMinus(sequence.ChildAt(i).Node));
+                    }
+#endif
+
+                    var last = sequence.ChildAt(sequence.ChildCount - 1).Node;
```

Comment:
I'm surprised. I expected we would need to know if we're on the left or the right side of a range. On the left, the character that matters is the last of the sequence, but on the right, it's the first one that matters.

Reply:
I will have to look at teh .net parser to see if we match them here.  note: this part of their parser is by far the most confusing (and likely buggy) bit.

Reply:
Ok.  I looked at the .net parser.   You can see the code for ScanCharClass here:

https://github.com/dotnet/corefx/blob/25c5878a7309b1361178e4eb3441af5b3e767c85/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L498

Here's my analysis of what's going on there (which i need to write tests for):

The main complexity is in: https://github.com/dotnet/corefx/blob/6ae0da1563e6e701bac61012c62ede8f8737f065/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L517

This is the loop that processes the character class a small chunk at a time, looking to see if:
1. the char class is done.
2. a char-class escape is found.
3. if we find a '-' and that starts a range.
4. if we find a '-' that starts a subtraction

Importantly, this loop is stateful (i.e. it updates function state as it moves through the character class), but it doesn't necessarily always update all state each time is moves.  For example, if you look at the switch that handles escapes (https://github.com/dotnet/corefx/blob/6ae0da1563e6e701bac61012c62ede8f8737f065/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L531), you will see that it assigns to 'ch', but then 'continues' back to the top of the loop.  That means, as we're processing each 'chunk' of hte character class, we are only recording the 'ch' for the last part we ran into.  

Similarly, once we hit a '-' that starts a character-class-range (https://github.com/dotnet/corefx/blob/6ae0da1563e6e701bac61012c62ede8f8737f065/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L632), we then store 'ch' into 'chPrev'.  This means that 'chPrev' is the value of the rightmost-chunk of hte left side. We then try to read in the chunks on the right, storing these value into 'ch'.  However, like above, we might hit escapes, causing us to continue, causing new values to be written into 'ch'.  

Finally, we'll break out of that loop and then see that we were on the right side of a range, and thus the range is stopped.  We'll then compare 'ch' to 'chPrev' (https://github.com/dotnet/corefx/blob/6ae0da1563e6e701bac61012c62ede8f8737f065/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L626).  'ch' will be the rightmost chunk of the *right* side of the range.  

So, in both cases, we're comparing the rightmost chunk of both sides of the range.  

Note: i don't believe any of this was intentionally thought through on the part of hte .net regex parser.  But it's the behavior they have, so it's the behavior i'm preserving.  I'll add tests for this now to validate that this is their behavior and that we match it accordingly.

Also tagging @danmosemsft for his assessment here.

Reply:
@viktorhofer actually owns regex; I just interfere.

Reply:
Note: i do already have tests for this.  For example: TestCharacterClassRange57

https://github.com/dotnet/roslyn/pull/25520/files#diff-c09d350e72f9ec3d676978c5c237e34bR13594

Here, we're testing the regex pattern: ```[b-\-a]```

in this case, due to how the .net regex engine works, the left side of the range is ```b```, and the right side of the range is ```\-a```.  IMO, this is almost certainly a bug in their impl (@danmosemsft  to confirm), but it means that based on how their impl works, they hit the ```\-``` then immediately continue the loop, instead of thinking that the range should be ```b-\-```.  So, we then hit ```a```, fall out of the loop, and check ```b``` against ```a```, reporting the issue that 'a' is less than 'b'.  

This is what teh .net parser does, and the behavior we replicate.

Reply:
Note: while i believe the existing .net regex parsing behavior is not intentional.  it's also the actual behavior that has shipped.  As such:

1. i'm uncertain if it could change.
2. it forces our hands as to the behavior we have.  Our goal is to mirror them, not be a proper implementation of the actual spec'ed documentation.

Reply:
To be brief. @danmosemsft i believe the following code is unintentional:

https://github.com/dotnet/corefx/blob/6ae0da1563e6e701bac61012c62ede8f8737f065/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L577-L580

If the .net regex parser encounters something like ```[#-\-b]``` it should stream this as the character class containing:  "teh range of characters from ```#``` to ```-```, along with the character ```b```".  instead, the .net parser will effectively ignore the ```\-``` on the right side of the range, and think this is "teh range of characters from ```#``` to ```b```".

Reply:
Sorry, will stop pinging you Dan.  @ViktorHofer ^

Reply:
Sorry I had an issue with GH notifications. Is this still relevant?

Reply:
@ViktorHofer Yes.  This looks like a bug in the .net native regex parser.  I wanted to check with you to confirm. My analysis is above, along with cases that seem to exhibit behavior that i feel could not reasonably be seen to be 'by design'.  I've still replicated this logic in the Roslyn parser.  But i would love to be able to remove these strange cases if possible if the .net native parser gets fixed up :)

Reply:
I believe there's no difference between the .NET Native and .NET Core parser, at least not code-wise.

Reply:
@ViktorHofer Sorry... i'm not sure what that means.  when i say ".net native" all i mean is "the real .net regex parser" as opposed to the "the new regex parser i created for roslyn".

The question for me is:  should we consider the behavior outlined in this conversation a bug in "the real .net regex parser"?  And shoudl Roslyn attempt to replicate it?   Or should we get the real .net parser to fix this behavior, and have Roslyn follow the new behavior?

Reply:
@ViktorHofer If you can, please read this entire subthread, and look at the pieces of code i've linked to and my explanations of the issue.  It very much seems like a bug in the current "real .net regex parser" impl to me. If you agree, i can file a bug on corefx to get this fixed.

Reply:
Ah I see. .NET Native = .NET Core flavor for UWP apps, .NET Framework = classic .NET.

Let me look into this later, I'll get back to you.

Reply:
This is definitely a bug and not by design. I would love this to be fixed but I fear that changing the existing behavior could break someone. I used a pattern searcher to go through some of our internal data and found 2 unique pattern which are built like that: `[.-\-.*]`.

Reply:
Ok correction, the two matches that I found aren't relevant here. It __could__ definitely be the case that nobody ever used the defected behavior...

Reply:
Ok cool.  If you do get around to fixing it, let me now so i can keep roslyn up to date.  Thanks!